src/pages/en/guides/content-collections.mdx

```diff
@@ -320,157 +232,73 @@ const blogPosts = await getCollection('blog', ({ data }) => {
+export async function getStaticPaths() {
+  const blog = await getCollection('blog');
+  return blog.map(entry => ({
+    params: { slug: entry.slug }, props: { entry },
+  }));
+}
 
-```astro "{ remarkPluginFrontmatter }"
----
-import { getCollection } from 'astro:content';
-const blogPosts = await getCollection('blog');
+const { entry } = Astro.props;
+const { Content } = await entry.render();
 ---
-
-{blogPosts.map(async (post) => {
-  const { remarkPluginFrontmatter } = await post.render();
-  return <p>{post.data.title} — {remarkPluginFrontmatter.readingTime}</p>
-})}
+<h1>{entry.data.title}</h1>
+<Content />
 ```
 
-Assuming `readingTime` was injected ([see our reading time example](/en/guides/markdown-content/#example-calculate-reading-time)), it will be available on the `remarkPluginFrontmatter` object.
-
-:::caution
-Remark and rehype plugins will have access to the _raw_ Markdown or MDX document frontmatter. This means frontmatter will not reflect any changes or defaults applied by your `schema`.
-:::
-
-<details>
-<summary>**🙋 Why don't `getCollection()` and `getEntry()` contain these values?**</summary>
-
-The remark and rehype pipelines are only run when your content is **rendered.** This lets `render()` access anything generated by these plugins like injected frontmatter. To stay performant, `getCollection()` and `getEntry()` do not have this capability.
-
-</details>
-
-## Generating pages from content collections
-
-You can create pages based on your content collections using [dynamic routes](/en/core-concepts/routing/#dynamic-routes). 
-
-Use `getCollection()` inside a [`getStaticPaths()`](/en/reference/api-reference/#getstaticpaths) function to query your content entries and provide the `slug` parameter for each page. 
-
-For example, you can dynamically create a page for each entry in a `blog` collection, including nested directories, by creating a `.astro` page with a [rest parameter in its filename](/en/core-concepts/routing/#rest-parameters) to match file paths of any depth.
-
-	```astro "{ slug: entry.slug }"
-	---
-	// src/pages/posts/[...slug].astro
-	import { getCollection } from 'astro:content';
-	
-	export async function getStaticPaths() {
-	  const blog = await getCollection('blog');
-	  return blog.map(entry => ({
-	    params: { slug: entry.slug },
-	  }));
-	}
-	---
-	```
-
 This will generate page routes for every entry in the `blog` collection, mapping each entry’s slug to a URL. For example, an entry at `src/content/blog/hello-world.md` will have a slug of `hello-world` and the collection entry `src/content/blog/en/intro.md` will have a slug of `en/intro`.
 
 Because this dynamic route is in `src/pages/posts/`, the final URLs will be `/posts/hello-world/` and `/posts/en/intro/`.
 
-### Rendering post contents
+#### Generating pages in SSR mode
```

Comment:
@bholmesdev I hate to be the bearer of bad news, but I think I found an issue with Content Collections when we switch to SSR. How does fetching an entry work in an SSR site, where you call `getEntry()` based on the `Astro.request.pathname` value?

`slug` is the wrinkle: if the pathname is my `slug`, but `getEntry()` is by `id`, you have no way to call `getEntry()`. `getEntry('blog', slugFromPathname)` wouldn't match the correct entry, as I understand it.

Lets chat on Tuesday, I have thoughts!

Reply:
No worries, very aware of this! We recommend using `getCollection()` with a filter today:

```ts
---
// taken from SSR tests in core
import { getCollection } from 'astro:content';

const { slug } = Astro.params;
const matches = await getCollection('blog', post => post.slug === slug);
const post = matches[0];
if (!post) {
	throw new Error(`Not found: ${slug}`);
}
const { Content } = await post.render();
---
```

I agree `getEntry()` could be changed to address this. Discussing with a user in our `#general` chat, turning `getEntry()` into a "find first match" function would be helpful here:

```diff
---
import { getEntry } from 'astro:content';

const { slug } = Astro.params;
- const matches = await getCollection('blog', post => post.slug === slug);
- const post = matches[0];
+ const post = await getEntry('blog', post => post.slug === slug);
...
```

So you can pass the ID string directly if known for an O(1) lookup time, or pass a function to lookup by other properties. This also unlocks lookup by `data` properties without a `matches` array.

We can also discuss ensuring all `slug`s are unique to allow direct lookup by slug:
```ts
const post = await getEntry('blog', { slug: Astro.params.slug });
```
...though this has more breaking implications, and wouldn't be as flexible. Since we bundle a map of the entire collection into your SSR build, regardless of whether you use `getEntry()` or `getCollection()`, I lean towards the first option. Happy to discuss further next week!

Reply:
Yes, let's definitely discuss early next week. But, I'll throw this idea out now so that its not a surprise on Tuesday, and hopefully having the weekend to sit with it helps:

- `slug` doubling as both nice-to-have meta-property _and_ an essential ID lookup is beginning to smell (for lack of a better term), especially since `id` is already meant to be that essential ID lookup property.
- `slug` may be redundant, given that we already have `id`
- `slug` can be removed, in favor of `id`

This was less of an issue up to this point when you consider `slug` as a nice-to-have built-in helper to help users create their URLs easier. Nothing wrong with that! But, given that "get by request path" will actually be an essential use-case for any `[slug].astro` page (SSG too, not just SSR), the confusion and redundancy between `id` and `slug` becomes more of a critical issue to resolve.

I see what you mean where this can be solved in userland by just using `getCollection()` and filtering. But I'd argue that getting the entire collection just to filter & grab a single entry is not a very friend API, and not what users would expect from such a basic use-case (we have feedback from users today to back this up, who are upset with having to do this with `import.meta.glob` and `Astro.glob`).

I also see what you mean where this could be added as an _additional_ feature to `getEntry()`, but that's where I'll ask us to take a step back and reconsider _why_ we need two different lookup methods, when really one should suffice?


### Alternative: Remove `slug` in favor of `id`

https://github.com/withastro/astro/compare/content-slug-to-id?expand=1

I spun up a quick prototype, merging `slug` and `id` together to create a more slug-like, unique `id`. Fun fact: this is actually something that I pushed for originally, back when `id` was the full file path. I think this presents an opportunity to solve a problem by revisiting this idea.


### Example

```diff
- <a href={'/blog/' + post.slug}>{post.data.title}</a>
+ <a href={'/blog/' + post.id}>{post.data.title}</a>
```

```diff
// src/pages/blog/index.astro
export async function getStaticPaths() {
  const posts = await getCollection('blog');
  return posts.map((post) => ({
-   params: { slug: post.slug },
+   params: { slug: post.id },
  }));
}
```

User code would be updated to use the entry ID instead of its slug. `id` becomes more slug-like to accommodate this. The main change to `id` is that it no longer has the file extension included. It also no longer uses github-slugger internally. Instead, it's expected that the user should use slug-friendly filenames in content collections.

```
src/content/blog/first-post.md         id: "first-post"
src/content/blog/foo/second-post.md    id: "foo/second-post"
src/content/blog/third post.md         warning/error
```

### Still to decide: can we support custom slugs/permalinks?

With slug gone, so is the current `slug()` customizer hook in `defineCollection`. I do love this idea of id/slug/permalink customization, and I'd still love to bring it back.

However, I'll reiterate here that this feature was never granted as an essential/required goal of the design doc, and has always been considered okay to skip for an initial release to be added later as a fast-follow after. IIRC, it was mainly seen as okay to add to initial release scope because it was seen as easy. I wouldn't want us to sacrifice on the critical APIs in an attempt to save a non-critical API that can be added later. 

One potentially easy answer would be to just take all the work already done on `slug()`, and rename it to `id()`. Dynamically changing IDs feels a bit awkward, but since that's already what the current proposal is doing with slugs I don't its a step backwards from where we are today and an acceptable compromise. Given that this is an advanced use-case, we would probably be safe to deprecate this later if we came up with something better.

I'd also like to explore an alternative proposal to support a static `id:` frontmatter property in any entry file, that manually sets the `id` (defaulting to the file name, if no `id:` frontmatter is given). I think the fact that this is what users already do and expect to work with the `permalink` property gives us a good datapoint that this is a pattern that users would enjoy. It's less powerful than the dynamic `slug()`/`id()` callback idea, but I think I could argue that the requirement of a `id:` frontmatter to be static and visible at the top of any file is easier to reason about.

### To be revisited Tuesday!

I'll be offline for the rest of the weekend dealing with house repair stuff, so I won't be able to revisit this or say much more until Tuesday. Given all the options that we have available, I'm confident that we can come up with a solution here next week without sacrificing our 2.0 launch timeline.

Reply:
I... disagree that custom `slug`s can be removed, since the RFC process and discussing with users both exposed custom `slug`s as important. I also don't see why reconciling `id` and `slug` _requires_ that custom `slug` mapping needs to be removed, since it fairly straightforward to ensure all generated slugs are unique at build time. I do agree that reconciling `id` and `slug` makes sense though. We can discuss then, yes!

Reply:
I ran into the SSR issue myself. My first thought was that `getEntry()` should return `| undefined` and that it would be your responsibility as a user to to convert your URL slug to a valid id. If you pass it a known static id it could omit the possible undefined. Not saying this is the solution, but it was my instinct.

----

I think we should be careful not to leak page concepts into content. Content should just be content. You can generate pages with the content in any number of ways.

That's why I don't like the "slug-like id" idea, @FredKSchott. This means you can't have `post.md` and `post.mdx` in the same collection. To me that violates the idea that it's content and we've just created another pages folder.

Reply:
Here's an idea that I think no one will like (probably even me!) but let me through it out anyways:

- Get rid of `getEntry()` entirely. How it's shown being used in the docs is probably not how anyone is actually going to use it. There might be times where you hard-code a known piece of content to get, but the confusion the API causes the minimal value created through the API.
- Get rid of the concept of `slug`. The user can add slug to their frontmatter themselves and use that as a key in this scenario, if they want.
  - This prevents the "duplicate id" problem.
  - This prevents leaking page concepts into content.
- Rename `id` to `filename` to more clearly identify what it is and to signal that the user is going to have to figure out a way to translate that to a slug (if they need to do so) or to use a frontmatter property of their own choosing.

Reply:
I think essential problem here is that we don't have the concept of a "page". A collection is generic. Which is good but makes this use-case harder.

I'd like to make a comparison to Wordpress, which does have the concept of a page. When you create a page it:

- Creates a permalink from the title
- Allows you to override the (permalink)

![5082c15022ac6e42b6b523e2f8acf0b1db9ecd0a](https://user-images.githubusercontent.com/361671/212937836-933747eb-e428-416d-8c9d-24a3b04c2f78.jpg)

So maybe this is the missing thing we need to add. Rather than making all collections of pages, make it so that you can signal which collections actually *are* used as pages.



One idea going off of this, in `defineCollection` add an optional `type` property whose value can be `'page'`. For collections of type pages you are *required* to provide a `slug`. You can then use a new querying API:

```astro
---
import { getPageBySlug } from 'astro:content';

const entry = getPageBySlug(Astro.params.slug);

if(!entry) return new Response(null, { status: 404 })
---
<!-- ... -->
```

Reply:
Hmm. What user issue does this extra complexity solve?

I want to make sure I understand the issue. The issue is that `getEntry` looks up by `id` but for “page” content, you more likely want to look up by `slug` as that’s what will be in `Astro.params` in SSR.

Main option today:

```js
const { slug } = Astro.params;
const [entry] = await getCollection('blog', (post) => post.slug === slug);
if (!entry) return new Response(null, { status: 404 });
```

A userland (or official) `getEntryBySlug(collection, slug)` abstraction around that is already possible today so is this really as big an issue as we think?

Reply:
Resolved in Discord, no change needed to `id`/`slug` but `getEntry` will be updated to `getEntryBySlug` to account for this much more common usecase.